src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,229 @@
+
+    elif invert_register == "sp_t":
+        new_circ = circuit.copy()
+
+    else:
+        raise NameError(
+            f"{invert_register} not recognized. Input "
+            "sp_0"
+            " to reset qubit 0, "
+            "sp_1"
+            " to reset qubit 1, or "
+            "sp_t"
+            " to reset both qubits."
+        )
+
+    inverse_circuit = new_circ.invert()
+
+    return inverse_circuit
+
+
+def GST_execute_circuit(circuit, k, j, nshots=int(1e4), backend=None):
+    """Executes a circuit used in gate set tomography and processes the
+        measurement outcomes for the Pauli Transfer Matrix notation. The circuit
+        should already have noise models implemented, if any, prior to using this
+        function.
+
+        Args:
+        circuit (:class:`qibo.models.Circuit`): The Qibo circuit to be executed.
+        k (int): The index of the state prepared.
+        j (int): The index of the measurement basis.
+        nshots (int, optional): Number of shots to execute circuit with.
+    Returns:
+        numpy.float: Expectation value given by either Tr(Q_j rho_k) or Tr(Q_j O_l rho_k).
+    """
+
+    nqubits = circuit.nqubits
+    if nqubits != 1 and nqubits != 2:
+        raise ValueError(
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+        )
+
+    else:
+        if j == 0:
+            return 1.0
+        else:
+            if backend is None:  # pragma: no cover
+                backend = GlobalBackend()
+
+            result = backend.execute_circuit(circuit, nshots=nshots)
+            observables = [symbols.I, symbols.Z, symbols.Z, symbols.Z]
+            observables_list = list(product(observables, repeat=nqubits))[j]
+            observable = 1
+            for q, obs in enumerate(observables_list):
+                if obs is not symbols.I:
+                    observable *= obs(q)
+            observable = SymbolicHamiltonian(observable, nqubits=nqubits)
+            expectation_val = result.expectation_from_samples(observable)
+            return expectation_val
+
+
+def GST(
+    nqubits=None,
+    gate=None,
+    nshots=int(1e4),
+    invert_register=None,
+    noise_model=None,
+    backend=None,
+):
+    """Runs gate set tomography for a 1 or 2 qubit gate/circuit.
+
+        Args:
+        nshots (int, optional): Number of shots used in Gate Set Tomography.
+        gate (:class:`qibo.gates.abstract.Gate`, optional): The gate to perform gate set tomography on.
+            If gate=None, then gate set tomography will be performed for an empty circuit.
+        noise_model (:class:`qibo.noise.NoiseModel`, optional): Noise model applied
+            to simulate noisy computation.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): Calculation engine.
+    Returns:
+        numpy.array: Numpy array with elements jk equivalent to either Tr(Q_j rho_k) or Tr(Q_j O_l rho_k).
```

Comment:
Since this is backend-agnostic, the return isn't always a `numpy.ndarray`

Reply:
```Tr(Q_j rho_k)``` or ```Tr(Q_j O_l rho_k)``` is given in a matrix form. If the return is not ```numpy.ndarray```, can we force a matrix / numpy array to be the output? 

It is not possible to proceed further with PEC if we don't have the matrix.

Reply:
The point is not being a matrix or not, it's that it can be a `cupy.ndarray` or a `tensorflow.EagerTensor`. So my suggestion (for now) is this:

```suggestion
        ndarray: array with elements ``jk`` equivalent to either :math:`\\text{tr}(Q_{j} \\, \\rho_{k})` or :math:`\\text{tr}(Q_{j} \\, O_{l} \\rho_{k})`.
```

Reply:
Ah ok, got it. Will change the description and include a definition for O_l.